marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -486,6 +492,109 @@ spec = do
              $ ensureMinUtxo protocolTestnet (inAddress, inValue)
          pure $ selectLovelace outValue `shouldBe` maximum [selectLovelace inValue, selectLovelace expected]
 
+  describe "balanceTx" do
+    prop "tx should balance if coin selection succeeded and no Plutus failure occurred" \(SomeTxConstraints marloweVersion constraints) -> do
+      marloweContext <- genSimpleMarloweContext marloweVersion constraints
+
+      -- We MUST dictate the distribution of wallet context assets, default
+      -- generation only tests with empty wallets!
+      maxLovelace <- choose (0, 40_000_000)
+      walletContext <- genWalletWithAsset marloweVersion constraints maxLovelace
+      start <- SystemStart <$> arbitrary
+
+      let
+        -- The following 4 definitions are for constructing a pure EraHistory,
+        -- which would normally come from the chain at runtime
+
+        eraHistory :: EraHistory CardanoMode
+        eraHistory = EraHistory CardanoMode
+          $ mkInterpreter
+          $ summaryWithExactly
+          $ Exactly
+          $  K (oneMillisecondEraSummary 0) -- Byron lasted 1 ms
+          :* K (oneMillisecondEraSummary 1) -- Shelley lasted 1 ms
+          :* K (oneMillisecondEraSummary 2) -- Allegra lasted 1 ms
+          :* K (oneMillisecondEraSummary 3) -- Mary lasted 1 ms
+          :* K (oneMillisecondEraSummary 4) -- Alonzo lasted 1 ms
+          :* K (unboundedEraSummary 5) -- Babbage never ends
+          :* Nil
+
+        unboundedEraSummary :: Integer -> EraSummary
+        unboundedEraSummary i = EraSummary
+          { eraStart = oneMillisecondBound i
+          , eraEnd = EraUnbounded
+          , eraParams = EraParams
+            { eraEpochSize = 1
+            , eraSlotLength = mkSlotLength 0.001
+            , eraSafeZone = UnsafeIndefiniteSafeZone
+            }
+          }
+
+        oneMillisecondEraSummary :: Integer -> EraSummary
+        oneMillisecondEraSummary i = EraSummary
+          { eraStart = oneMillisecondBound i
+          , eraEnd = EraEnd $ oneMillisecondBound $ i + 1
+          , eraParams = EraParams
+            { eraEpochSize = 1
+            , eraSlotLength = mkSlotLength 0.001
+            , eraSafeZone = UnsafeIndefiniteSafeZone
+            }
+          }
+
+        oneMillisecondBound :: Integer -> Bound
+        oneMillisecondBound i = Bound
+          { boundTime = RelativeTime $ fromInteger i / 1000
+          , boundSlot = fromInteger i
+          , boundEpoch = fromInteger i
+          }
+
+        -- We need to make a TxBodyContent that would have come from executing
+        -- selectCoins, containing the tx information in the WalletContext we
+        -- will also be passing to balanceTx. To do so, we'll use the
+        -- walletContext.
+
+        addBuilder :: TxIn -> (TxIn, BuildTxWith BuildTx (Witness WitCtxTxIn BabbageEra))
+        addBuilder = (, BuildTxWith (KeyWitness KeyWitnessForSpending))
+
+        txBodyContent = emptyTxBodyContent
```

Comment:
This `txBodyContent` just spends the whole wallet. It ignores the `marloweContext` and it doesn't create a transaction that matches the description of this property, "tx should balance if coin selection succeeded and no Plutus failure occurred", since coin selection was never called and there is no possibility of the Marlowe validator being run. Perhaps the test should be renamed "tx should balance for non-Plutus transactions where the wallet has sufficient funds"?

Reply:
I was using `selectCoins` initially but reworked it to not because @jhbertra was concerned about potentially buggy code being involved in this test, which is valid.
I don't understand how to make a wallet + marlowe context + tx body that looks close to what comes out of successful coin selection.

Reply:
This is partly my fault - I discouraged @dino-iog from calling `selectCoins` in the test, but you are right that the test title indicates that it should be.

Reply:
Perhaps we should just rename the test?